File: apps/web/lib/qip/collections.ts
/**
 * QIP — Faz 5 · Sınav Koleksiyonları (BANK_QUESTİON Part 16).
 *
 * Bankadan OTOMATİK, deterministik koleksiyonlar üretir: Günün/Haftanın Sınavı, Başlangıç, Zor
 * Sorular, Yalnız İşaretler/Motor/İlk Yardım, En Çok Yanılan, Rastgele 50, AI Challenge.
 * Tohum (gün/hafta) UI'dan verilir → gün içinde sabit, günden güne farklı. Faz 2 kalite/tema +
 * Faz 3 aileleri + Faz 5 dinamik sınav üreticisini kullanır.
 */
import { hash32 } from '@ea/content-schema';
import { analyzedQuestions, type AnalyzedQuestion } from './index';
import { buildDynamicExam } from './exam';
import { seededRng, type Rng } from './visual';
import type { QuestionAnalytics } from './analytics';

export interface CollectionSpec {
  id: string;
  label: string;
  description: string;
  emoji: string;
  count: number;
  questionIds: string[];
}

export interface CollectionsOptions {
  /** Gün tohumu (ör. 'YYYY-MM-DD' → hash). Verilmezse sabit 1. */
  daySeed?: number;
  weekSeed?: number;
  /** En-çok-yanılan sıralaması için opsiyonel soru istatistikleri. */
  stats?: QuestionAnalytics[];
  pool?: AnalyzedQuestion[];
}

/** Tarih dizesinden kararlı tohum (UI: `seedFromDate('2026-07-21')`). */
export function seedFromDate(dateStr: string): number {
  return parseInt(hash32(dateStr), 16) >>> 0;
}

/** ISO hafta anahtarı için basit hafta tohumu (aynı haftada sabit). */
export function seedFromWeek(dateStr: string): number {
  // YYYY-MM-DD → yıl + haftanın kabaca indeksini kullan (deterministik, UI'dan verilir).
  return parseInt(hash32('week:' + dateStr.slice(0, 7)), 16) >>> 0;
}

function pick(pool: AnalyzedQuestion[], rng: Rng, n: number): string[] {
  const a = [...pool];
  for (let i = a.length - 1; i > 0; i--) {
    const j = Math.floor(rng() * (i + 1));
    [a[i], a[j]] = [a[j]!, a[i]!];
  }
  return a.slice(0, n).map((q) => q.id);
}

/** Otomatik koleksiyonları üret (GERÇEK sayılar). */
export function examCollections(opts: CollectionsOptions = {}): CollectionSpec[] {
  const pool = opts.pool ?? analyzedQuestions();
  const daySeed = opts.daySeed ?? 1;
  const weekSeed = opts.weekSeed ?? 2;

  const out: CollectionSpec[] = [];
  const add = (id: string, label: string, description: string, emoji: string, ids: string[]) =>
    out.push({ id, label, description, emoji, count: ids.length, questionIds: ids });

  // Günün / Haftanın Sınavı — dinamik, dengeli, benzersiz.
  add(
    'gunun-sinavi',
    'Günün Sınavı',
    'Her gün yenilenen, MEB dağılımına uygun dengeli 50 soruluk sınav.',
    '📅',
    buildDynamicExam({ seed: daySeed, count: 50 }).questions.map((q) => q.id)
  );
  add(
    'haftanin-sinavi',
    'Haftanın Sınavı',
    'Haftalık, zorluğu dengelenmiş 50 soruluk sınav.',
    '🗓️',
    buildDynamicExam({ seed: weekSeed, count: 50, difficultyBalance: true }).questions.map(
      (q) => q.id
    )
  );

  // Başlangıç — kolay sorular.
  add(
    'baslangic',
    'Başlangıç',
    'Yeni başlayanlar için kolay sorular.',
    '🌱',
    pick(
      pool.filter((q) => q.difficulty === 'kolay'),
      seededRng(daySeed + 11),
      40
    )
  );

  // Zor Sorular.
  add(
    'zor-sorular',
    'Zor Sorular',
    'Kendini sına: yalnızca zor sorular.',
    '🔥',
    pick(
      pool.filter((q) => q.difficulty === 'zor'),
      seededRng(daySeed + 22),
      40
    )
  );

  // Yalnız İşaretler.
  add(
    'sadece-isaretler',
    'Yalnız Trafik İşaretleri',
    'Trafik işaretleri temalı sorular.',
    '🚦',
    pick(
      pool.filter((q) => q.primaryTheme === 'trafik-isaretleri' || q.topic === 'isaretler'),
      seededRng(daySeed + 33),
      40
    )
  );

  // Yalnız Motor.
  add(
    'sadece-motor',
    'Yalnız Araç Tekniği',
    'Motor ve araç tekniği soruları.',
    '🔧',
    pick(
      pool.filter((q) => q.subject === 'motor'),
      seededRng(daySeed + 44),
      40
    )
  );

  // Yalnız İlk Yardım.
  add(
    'sadece-ilkyardim',
    'Yalnız İlk Yardım',
    'İlk yardım bilgisi soruları.',
    '🚑',
    pick(
      pool.filter((q) => q.subject === 'ilkyardim'),
      seededRng(daySeed + 55),
      40
    )
  );

  // En Çok Yanılan — istatistik varsa başarısızlık oranına göre; yoksa zorluk+süre.
  const mostFailed = mostFailedIds(pool, opts.stats, 40);
  add(
    'en-cok-yanilan',
    'En Çok Yanılan',
    opts.stats?.length
      ? 'Kullanıcıların en çok yanıldığı sorular.'
      : 'En zorlayıcı sorular (zorluk + tahmini süreye göre).',
    '⚠️',
    mostFailed
  );

  // Rastgele 50.
  add(
    'rastgele-50',
    'Rastgele 50',
    'Bankadan rastgele 50 soru.',
    '🎲',
    pick(pool, seededRng(daySeed + 66), 50)
  );

  // AI Challenge — en zor + görsel + çok-varyantlı karışımı.
  const challenge = pick(
    pool.filter((q) => q.difficulty === 'zor' || q.image || (q.qualityScore ?? 100) < 100),
    seededRng(daySeed + 77),
    30
  );
  add(
    'ai-challenge',
    'AI Challenge',
    'Zorluğu yüksek, karışık meydan okuma seti.',
    '🤖',
    challenge
  );

  return out;
}

/** En çok yanılan id'ler: istatistik varsa yanlış oranına göre; yoksa zorluk+süre yaklaşığı. */
function mostFailedIds(
  pool: AnalyzedQuestion[],
  stats: QuestionAnalytics[] | undefined,
  n: number
): string[] {
  if (stats && stats.length) {
    const byId = new Map(stats.map((s) => [s.questionId, s]));
    return [...pool]
      .filter((q) => byId.has(q.id) && (byId.get(q.id)!.attempts ?? 0) >= 3)
      .sort((a, b) => (byId.get(b.id)!.wrongRate ?? 0) - (byId.get(a.id)!.wrongRate ?? 0))
      .slice(0, n)
      .map((q) => q.id);
  }
  const rank = (q: AnalyzedQuestion) =>
    (q.difficulty === 'zor' ? 2 : q.difficulty === 'orta' ? 1 : 0) * 100 + q.estimatedSeconds;
  return [...pool]
    .sort((a, b) => rank(b) - rank(a))
    .slice(0, n)
    .map((q) => q.id);
}

/** id → koleksiyon çözümü (UI). */
export function collectionById(id: string, opts?: CollectionsOptions): CollectionSpec | undefined {
  return examCollections(opts).find((c) => c.id === id);
}

File: apps/mobile/lib/domain/practice/collections.dart
import '../onboarding/study_profile.dart';
import 'exam.dart';
import 'question.dart';

/// Otomatik sınav koleksiyonları — web `qip/collections.ts`'in mobil (analiz-katmansız) uyarlaması.
/// Deterministik: gün/hafta tohumundan → gün içinde sabit, günden güne farklı. Temalı setler bankadan
/// tohumlu seçilir (analiz katmanı yerine doğrudan alan filtreleri; pratik seti için yeterli).

class CollectionSpec {
  const CollectionSpec({
    required this.id,
    required this.label,
    required this.description,
    required this.emoji,
    required this.questionIds,
  });
  final String id;
  final String label;
  final String description;
  final String emoji;
  final List<String> questionIds;

  int get count => questionIds.length;
}

List<String> _pick(List<Question> pool, Rng rng, int n) =>
    shuffle(pool, rng).take(n).map((q) => q.id).toList();

/// Evolution Faz E5 — sınıfa özgü odak seti için kavram örüntüleri.
///
/// ÖNEMLİ: bu setlerin soruları BANKADAN gelir; hiçbir soru uydurulmaz veya sınıfa göre yeniden
/// yazılmaz. e-Sınav teori bankası tüm sınıflarda ORTAKTIR; burada yapılan tek şey, o sınıfın
/// aracını/mesleğini doğrudan konu alan gerçek soruları bir araya getirmektir.
final RegExp _motoPattern = RegExp(
  r'motosiklet|motorlu bisiklet|kask(?!o)|skuter',
  caseSensitive: false,
);
final RegExp _busPattern = RegExp(
  r'otobüs|minibüs|yolcu taşı|yolcu indir|yolcu bindir|ticari araç|kamyon|çekici|römork|taşıt katarı|okul taşıt|servis araç',
  caseSensitive: false,
);

/// Sorunun tüm metni (kök + şıklar + açıklama + konu) — eşleşme bunun üzerinde yapılır.
String _haystack(Question q) => [q.stem, ...q.options, q.explanation, q.topic].join(' ');

/// Bu sınıfın aracını doğrudan konu alan bankadaki gerçek sorular.
List<Question> licenceFocusQuestions(List<Question> bank, LicenceCategory licence) {
  final pattern = switch (licence) {
    LicenceCategory.a => _motoPattern,
    LicenceCategory.d => _busPattern,
    // B sınıfı bankanın varsayılan odağıdır; yapay bir alt küme üretmek yanıltıcı olur.
    LicenceCategory.b => null,
  };
  if (pattern == null) return const [];
  return bank.where((q) => pattern.hasMatch(_haystack(q))).toList();
}

/// Koleksiyonları üret (gerçek sayılar). [daySeed]/[weekSeed] tarih tohumlarıdır.
/// [licence] verilirse sınıfa özgü odak seti listenin BAŞINA eklenir (Faz E5).
List<CollectionSpec> examCollections(
  List<Question> bank, {
  required int daySeed,
  required int weekSeed,
  LicenceCategory? licence,
}) {
  final out = <CollectionSpec>[];
  void add(String id, String label, String description, String emoji, List<String> ids) {
    out.add(
      CollectionSpec(id: id, label: label, description: description, emoji: emoji, questionIds: ids),
    );
  }

  if (licence != null) {
    final focus = licenceFocusQuestions(bank, licence);
    if (focus.isNotEmpty) {
      // Tümü alınır (kırpılmaz) → kartta görünen sayı gerçek sayıdır; sıra günlük tohumla karışır.
      final ids = shuffle(focus, seededRng(daySeed + 77)).map((q) => q.id).toList();
      switch (licence) {
        case LicenceCategory.a:
          add(
            'motosiklet-odakli',
            'Motosiklet Odaklı',
            'Bankadaki motosiklet, kask ve iki tekerlek konulu gerçek sorular.',
            '🏍️',
            ids,
          );
        case LicenceCategory.d:
          add(
            'otobus-odakli',
            'Otobüs ve Yolcu Taşımacılığı',
            'Bankadaki otobüs, yolcu taşımacılığı ve ağır araç konulu gerçek sorular.',
            '🚌',
            ids,
          );
        case LicenceCategory.b:
          break;
      }
    }
  }

  // Günün / Haftanın Sınavı — MEB dağılımına uygun dengeli 50 soru (tohumlu buildExam).
  add(
    'gunun-sinavi',
    'Günün Sınavı',
    'Her gün yenilenen, MEB dağılımına uygun dengeli 50 soruluk sınav.',
    '📅',
    buildExam(bank, rng: seededRng(daySeed)).questions.map((q) => q.id).toList(),
  );
  add(
    'haftanin-sinavi',
    'Haftanın Sınavı',
    'Haftalık, MEB dağılımına uygun 50 soruluk sınav.',
    '🗓️',
    buildExam(bank, rng: seededRng(weekSeed)).questions.map((q) => q.id).toList(),
  );

  add(
    'baslangic',
    'Başlangıç',
    'Yeni başlayanlar için kolay sorular.',
    '🌱',
    _pick(bank.where((q) => q.difficulty == Difficulty.kolay).toList(), seededRng(daySeed + 11), 40),
  );
  add(
    'zor-sorular',
    'Zor Sorular',
    'Kendini sına: yalnızca zor sorular.',
    '🔥',
    _pick(bank.where((q) => q.difficulty == Difficulty.zor).toList(), seededRng(daySeed + 22), 40),
  );
  add(
    'sadece-isaretler',
    'Yalnız Trafik İşaretleri',
    'Trafik işaretleri temalı sorular.',
    '🚦',
    _pick(bank.where((q) => q.topic.contains('isaret')).toList(), seededRng(daySeed + 33), 40),
  );
  add(
    'sadece-motor',
    'Yalnız Araç Tekniği',
    'Motor ve araç tekniği soruları.',
    '🔧',
    _pick(bank.where((q) => q.subject.name == 'motor').toList(), seededRng(daySeed + 44), 40),
  );
  add(
    'sadece-ilkyardim',
    'Yalnız İlk Yardım',
    'İlk yardım bilgisi soruları.',
    '🚑',
    _pick(bank.where((q) => q.subject.name == 'ilkyardim').toList(), seededRng(daySeed + 55), 40),
  );
  add(
    'en-cok-yanilan',
    'En Çok Yanılan',
    'En zorlayıcı sorular (zorluğa göre).',
    '⚠️',
    _pick(
      bank.where((q) => q.difficulty != Difficulty.kolay).toList(),
      seededRng(daySeed + 88),
      40,
    ),
  );
  add(
    'rastgele-50',
    'Rastgele 50',
    'Bankadan rastgele 50 soru.',
    '🎲',
    _pick(bank, seededRng(daySeed + 66), 50),
  );

  // Boş kalanları (banka bir temada yetersizse) at.
  return out.where((c) => c.questionIds.isNotEmpty).toList();
}

CollectionSpec? collectionById(
  List<Question> bank,
  String id, {
  required int daySeed,
  required int weekSeed,
  LicenceCategory? licence,
}) {
  for (final c in examCollections(
    bank,
    daySeed: daySeed,
    weekSeed: weekSeed,
    licence: licence,
  )) {
    if (c.id == id) return c;
  }
  return null;
}
